File: 016_3Sum_Closest/016_1.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>
#include <math.h>
#include <stdint.h>
#include <string.h>

int abs(x) {
    return x >= 0 ? x : -x;
}

int comp(const void *p1, const void *p2) {
    return *(const int *)p1 - *(const int *)p2;
}

int threeSumClosest(int* nums, int numsSize, int target) {
    qsort(nums, numsSize, sizeof(int), comp);
    int res = nums[0] + nums[1] + nums[2];
    for (int i = 0; i < numsSize - 2; i++) {
        int j = i + 1, k = numsSize - 1;
        while (j < k) {
            int cur = nums[i] + nums[j] + nums[k];
            if (abs(cur - target) < abs(res - target)) res = cur;
            if (cur == target) return res;
            else if (cur < target) j++;
            else if (cur > target) k--;
        }
    }
    return res;
}

int main(void) {
    int nums[] = {-1, 2, 1, -4};
    printf("%d\n", threeSumClosest(nums, 4, 1));
}

File: 016_3Sum_Closest/016_1.py
from typing import List, Dict
from queue import Queue
from collections import deque, Counter
import math
from heapq import heappush, heappushpop

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = nums[0] + nums[1] + nums[2]
        length = len(nums)
        for i in range(length - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            j, k = i + 1, length - 1
            while j < k:
                cur = nums[i] + nums[j] + nums[k]
                if abs(cur - target) < abs(res - target):
                    res = cur
                if cur == target:
                    return res
                elif cur < target:
                    j += 1
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                elif cur > target:
                    k -= 1
                    while j < k and nums[k] == nums[k+1]:
                        k -= 1
        return res